### PR TITLE
Runtime Assumptions infrastructure for Read Only Code

### PR DESCRIPTION
**Pull request by dsouzai**
https://github.com/eclipse/omr/issues/5542

This PR adds some infrastructure changes to the runtime assumptions to facilitate support for a read only code cache. Specifically, 

1. `serialize` and `size` APIs are added, which child classes of `OMR::RuntimeAssumption` can extend
2. A pointer to the owning metadata is added to the Sentinel Assumption; this allows one to dangle a chain of runtime assumptions associated with a specific compiled body off of a metadata structure that describes said body when reifying the assumptions is necessary (e.g. in a snapshot/restore implementation).

### CODE REVIEW

**Comment by dsouzai** [comment]
@0xdaryl could you please review.

**Comment by 0xdaryl** [comment]
Code changes themselves look fine, but could you add Doxygen comments for the new APIs please?

**Comment by dsouzai** [comment]
@0xdaryl added doxygen.

File: compiler/runtime/OMRRuntimeAssumptions.hpp [hunk 1]
```diff
@@ -268,7 +268,13 @@ class SentinelRuntimeAssumption : public OMR::RuntimeAssumption
...
    virtual uint8_t *getFirstAssumingPC() { return NULL; }
    virtual uint8_t *getLastAssumingPC() { return NULL; }
-   virtual void     dumpInfo() {};
+   virtual void     dumpInfo() {}
+
+   void *getOwningMetadata()                     { return _owningMetaData; }
+   void  setOwningMetadata(void *owningMetadata) { _owningMetaData = owningMetadata; }
+
+   private:
+   void * _owningMetaData;
```
Comment (0xdaryl) [inline compiler/runtime/OMRRuntimeAssumptions.hpp:291]: I like the comment you have in the commit message for this field.  Why not add it as a code comment here describing the purpose of this field:

```
A pointer to the owning metadata.  This allows one to dangle a chain of runtime assumptions 
associated with a specific compiled body off of a metadata structure that describes said 
body when reifying the assumptions is necessary.
```

Comment (dsouzai) [reply compiler/runtime/OMRRuntimeAssumptions.hpp:296]: Done in https://github.com/eclipse/omr/pull/5553/commits/90b975bf9cdcd13acdae2ec7e550e8aa01327b83

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**